### PR TITLE
Use standard Julia mechanisms for LibGit2 StrArrayStruct and Buffer types.

### PR DESCRIPTION
**Pull request by simonbyrne**
The code currently calls `Libc.malloc`, and then relies on the library
function to free the data. This seems slightly dubious, so I rejigged it
to use `cconvert` and `unsafe_convert`.

### CODE REVIEW

File: base/libgit2/strarray.jl [hunk 1]
```diff
@@ -1,37 +1,15 @@
...
-StrArrayStruct{T<:AbstractString}(strs::Vector{T}) = StrArrayStruct(strs...)
-
-function Base.convert(::Type{Vector{AbstractString}}, sa::StrArrayStruct)
-    arr = Array{AbstractString}(sa.count)
-    for i=1:sa.count
-        arr[i] = unsafe_string(unsafe_load(sa.strings, i))
-    end
-    return arr
+function Base.unsafe_convert(::Type{Ptr{StrArrayStruct}}, tup::Tuple)
+    Base.unsafe_convert(Ptr{StrArrayStruct}, first(tup))
```
Comment (tkelman) [inline base/libgit2/strarray.jl]: why does this throw away all but the first element?

File: base/libgit2/types.jl [hunk 2]
```diff
@@ -38,34 +38,61 @@ end
...
+When fetching data from LibGit2, a typical usage would look like:
+```
+sa_ref = Ref(StrArrayStruct())
+@check ccall(..., (Ptr{StrArrayStruct},), sa_ref)
+res = Base.unsafe_convert(Vector{String}, sa_ref[])
+free(sa_ref)
+```
+In particular, note that `LibGit2.free` should be called afterward on the `Ref` object.
+
+Conversely, when a vector of strings to LibGit2, it is generally simplest to rely on
```
Comment (tkelman) [inline base/libgit2/types.jl]: when passing?

File: base/libgit2/strarray.jl [hunk 3]
```diff
@@ -1,37 +1,15 @@
...
-        unsafe_copy!(out_ptr, in_ptr, len)
-        unsafe_store!(out_ptr, zero(UInt8), len + 1) # NULL byte
-        unsafe_store!(sa_strings, convert(Cstring, out_ptr), i)
-    end
-    return StrArrayStruct(sa_strings, strcount)
+function Base.cconvert{T<:AbstractString}(::Type{Ptr{StrArrayStruct}}, strs::Vector{T})
+    data = [Base.cconvert(Cstring, s) for s in strs]
+    ptrs = [Base.unsafe_convert(Cstring, d) for d in data]
+    sa = Ref(StrArrayStruct(pointer(ptrs), length(ptrs)))
+    sa, data, ptrs
```
Comment (simonbyrne) [inline base/libgit2/strarray.jl]: @tkelman related to your comment below: the idea is that `cconvert` returns a tuple but maintains a reference to the original data.

I agree it's not very pretty: any ideas for a better way to do it?

File: base/libgit2/diff.jl [hunk 4]
```diff
@@ -1,24 +1,29 @@
...
+    do_ref = Ref(DiffOptions(pathspec = sa))
+    return do_ref, data, cs
+end
+function Base.unsafe_convert(::Type{Ptr{DiffOptionsStruct}}, tup::Tuple)
+    Base.unsafe_convert(Ptr{DiffOptionStruct}, first(tup))
+end
+
+
+
+
```
Comment (ararslan) [inline base/libgit2/diff.jl]: Really pedantic point, but perhaps 4 line breaks is an unnecessary amount of vertical space?

File: base/libgit2/types.jl [hunk 5]
```diff
@@ -38,34 +38,61 @@ end
...
     LibGit2.StrArrayStruct
 
-Array of strings.
+A LibGit2 representation of an array of strings.
 Matches the [`git_strarray`](https://libgit2.github.com/libgit2/#HEAD/type/git_strarray) struct.
-"""
-@kwdef immutable StrArrayStruct
+
+When fetching data from LibGit2, a typical usage would look like:
+```
```
Comment (ararslan) [inline base/libgit2/types.jl]: \```julia?

File: base/libgit2/strarray.jl [hunk 6]
```diff
@@ -1,37 +1,15 @@
...
-StrArrayStruct{T<:AbstractString}(strs::Vector{T}) = StrArrayStruct(strs...)
 
-function Base.convert(::Type{Vector{AbstractString}}, sa::StrArrayStruct)
-    arr = Array{AbstractString}(sa.count)
-    for i=1:sa.count
-        arr[i] = unsafe_string(unsafe_load(sa.strings, i))
-    end
-    return arr
+function Base.cconvert(::Type{Ptr{StrArrayStruct}}, x::Vector)
+    str_ref = Base.cconvert(Ref{Cstring}, x)
```
Comment (simonbyrne) [inline base/libgit2/strarray.jl]: @vtjnash Does this look okay?

File: base/libgit2/strarray.jl [hunk 7]
```diff
@@ -1,37 +1,15 @@
...
+    Base.unsafe_convert(Ptr{StrArrayStruct}, first(rr))
 end
 
-function Base.copy(src::StrArrayStruct)
-    dst_ptr = Ref(StrArrayStruct())
-    @check ccall((:git_strarray_copy, :libgit2), Cint,
-                  (Ptr{StrArrayStruct}, Ptr{StrArrayStruct}),
-                   dst_ptr, Ref(src))
-    return dst_ptr[]
+function Base.unsafe_convert(::Type{Vector{String}}, sa::StrArrayStruct)
```
Comment (vtjnash) [inline base/libgit2/strarray.jl]: I think this is simply `convert`, no `unsafe_` prefix necessary.

File: base/libgit2/strarray.jl (hunk 6 above)
Comment (vtjnash) [inline base/libgit2/strarray.jl]: yes

**Comment by simonbyrne** [comment]
Okay, I'm fairly happy with this now. Thanks everyone.

**Pull request merged**